com.ca.lsp.cobol/lsp-service-cobol/src/main/java/com/ca/lsp/cobol/service/CobolWorkspaceServiceImpl.java

```diff
@@ -67,66 +76,72 @@ public CobolWorkspaceServiceImpl(DataBusBroker dataBus, Provider<LanguageClient>
   @Nonnull
   @Override
   public CompletableFuture<Object> executeCommand(@Nonnull ExecuteCommandParams params) {
+    runAsync(executeCopybookFix(params)).whenComplete(reportExceptionIfFound(params));
+
     return completedFuture(null);
   }
 
+  private Runnable executeCopybookFix(@Nonnull ExecuteCommandParams params) {
+    return () -> {
+      if (MISSING_COPYBOOK.name().equals(params.getCommand())) {
+        rerunAnalysis();
+      }
+    };
+  }
+
   /**
-   * This is a notification triggered automatically when the user modify configuration settings in
-   * the client
+   * Process changed configuration on the client state. This notification triggered automatically
+   * when the user modify configuration settings in the client. Invalidate all the caches to avoid
+   * dirty state.
    *
    * @param params - LSPSpecification -> The actual changed settings; Actually -> null all the time.
    */
   @Override
   public void didChangeConfiguration(DidChangeConfigurationParams params) {
-    try {
-
-      // invalidate cache to avoid false positive
-      dataBus.invalidateCache();
-      fetchSettings(LSP_PREFIX.label + "." + CPY_MANAGER.label, null)
-          .thenAccept(e -> dataBus.postData(FetchedSettingsEvent.builder().content(e).build()));
-    } catch (RuntimeException e) {
-      log.error(e.getMessage());
-    }
+    settingsService
+        .getConfiguration(LOCAL_PATHS.label)
+        .thenAccept(it -> acceptSettingsChange(toStrings(it)));
   }
 
-  /**
-   * After client notifies the server that there is a setting change we need to request the client
-   * those changing by sending a workspace/configuration JSON request
-   *
-   * @param section - The configuration section asked for.
-   * @param scope - The scope to get the configuration section for.
-   * @return - CompletedFuture which contains an object with the settings asked for.
-   */
-  private CompletableFuture<List<Object>> fetchSettings(String section, String scope) {
-    ConfigurationParams params =
-        new ConfigurationParams(provideConfigurationItemList(section, scope));
-    return clientProvider.get().configuration(params);
+  private void acceptSettingsChange(List<String> localFolders) {
```

Comment:
The method called asynchronously. Inside we add and remove watchers. It could cause us an issue if this method will be called in two threads. Is it make sense to use a synchronized method here?

Reply:
I think it is better to mark `WatcherServiceImpl#addWatchers` and `WatcherServiceImpl#removeWatchers` as `synchronized` to reduce the sync area